frame/evm/src/precompiles.rs

```diff
@@ -165,3 +178,273 @@ impl Precompile for Sha256 {
+
+		// Gas formula allows arbitrary large exp_len when base and modulus are empty, so we need to handle empty base first.
+		let r = if base_len == 0 && mod_len == 0 {
+			BigUint::zero()
+		} else {
+			// read the numbers themselves.
+			let mut buf = Vec::with_capacity(max(mod_len, max(base_len, exp_len)));
+			buf.copy_from_slice(&input[0..base_len]);
+			let base = BigUint::from_bytes_be(&buf[..base_len]);
+
+			buf = Vec::with_capacity(max(mod_len, max(base_len, exp_len)));
+			buf.copy_from_slice(&input[base_len..base_len + exp_len]);
+			let exponent = BigUint::from_bytes_be(&buf[..exp_len]);
+
+			buf = Vec::with_capacity(max(mod_len, max(base_len, exp_len)));
+			buf.copy_from_slice(&input[(base_len + exp_len)..(base_len + exp_len + mod_len)]);
+			let modulus = BigUint::from_bytes_be(&buf[..mod_len]);
+
+			if modulus.is_zero() || modulus.is_one() {
+				BigUint::zero()
+			} else {
+				base.modpow(&exponent, &modulus)
+			}
+		};
+
+		// write output to given memory, left padded and same length as the modulus.
+		let bytes = r.to_bytes_be();
+
+		// always true except in the case of zero-length modulus, which leads to
+		// output of length and value 1.
+		if bytes.len() <= mod_len {
+			let res_start = mod_len - bytes.len();
+			let mut ret = Vec::with_capacity(bytes.len() - mod_len);
+			ret.copy_from_slice(&bytes[res_start..bytes.len()]);
+			Ok((ExitSucceed::Returned, ret.to_vec(), cost))
+		} else {
+			Err(ExitError::Other("failed"))
+		}
+	}
+}
+
+#[cfg(feature = "ed25519")]
+pub struct Ed25519Verify;
+
+#[cfg(feature = "ed25519")]
+impl Precompile for Ed25519Verify {
+	fn execute(
+		input: &[u8],
+		target_gas: Option<usize>,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError> {
+		let cost = ensure_linear_cost(target_gas, input.len(), 15, 3)?;
+
+		let len = min(input.len(), 128);
+
+		let mut i = [0u8; 128];
+		i[..len].copy_from_slice(&input[..len]);
+
+		let mut buf = [0u8; 4];
+
+		let msg = &i[0..32];
+		let pk = PublicKey::from_bytes(&i[32..64])
+			.map_err(|_| ExitError::Other("Public key recover failed"))?;
+		let sig = Signature::try_from(&i[64..128])
+			.map_err(|_| ExitError::Other("Signature recover failed"))?;
+
+		// https://docs.rs/rust-crypto/0.2.36/crypto/ed25519/fn.verify.html
+		if pk.verify(msg, &sig).is_ok() {
+			buf[3] = 0u8;
+		} else {
+			buf[3] = 1u8;
+		};
+
+		Ok((ExitSucceed::Returned, buf.to_vec(), cost))
+	}
+}
+
+/// The Blake2F precompile.
+#[cfg(feature = "blake2")]
+pub struct Blake2F;
+
+#[cfg(feature = "blake2")]
+impl Precompile for Blake2F {
+	fn execute(
+		input: &[u8],
+		target_gas: Option<usize>,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError> {
+		let cost = ensure_linear_cost(target_gas, input.len(), 15, 3)?;
+        let mut state = Blake2b::new(64);
+        state.update(&input);
+        let ret = state.finalize().as_bytes().to_vec();
+		Ok((ExitSucceed::Returned, ret, cost))
+	}
+}
```

Comment:
This doesn't seem to be the [EIP-152 precompile](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-152.md), but rather the full Blake2 hash function.